### ISSUE TITLE
cache reporting as full but it isn't according to metrics

### ISSUE THREAD

**Issue by goir** [unbit/uwsgi#1168]
Hi,

we're having some problems with full caches. The uwsgi log states they are full but clearly aren't according to uwsgicachetop and the raw metrics.

	{
		"name":"XXXX",
		"hash":"djb33x",
		"hashsize":65536,
		"keysize":57,
		"max_items":30000,
		"blocks":30000,
		"blocksize":5,
		"items":24139,
		"hits":338199,
		"miss":26053,
		"full":25964,
		"last_modified_at":0
	},
`cache2 = name=XXXX,items=30000,keysize=57,blocksize=5,store_delete=1,store=%(cache_dir)/XXXX,ignore_full=1,purge_lru=1`

from the startup log:
*** Cache "XXXX" initialized: 4MB (key: 145 bytes, keys: 4350000 bytes, data: 150000 bytes, bitmap: 0 bytes) preallocated ***

this one is configured as lru cache but it should use the full 30k items. This also happens when not using lru. One the cache is full the item number does not change at all, it always stays at 24139. Items in this cache have a TTL of 1800 seconds.

If i do a reload of the app the number of items changes: `"items":24181`, `"items":24171`. But the cache is instantly full and the number does not change.

We've had another cache which said it was full at 64k items but is configured for 300k items. I've deleted the store file and restarted the app. It went over 100k items without being full again. This happened a few times now and it will again.
I think this has something to do with the store files which might get corrupted or something over time. Maybe restarting while the files are written ?

Is there a way to analyze the store file if this happens again or any other idea how to find what the issue is ?

Thanks
Goir

**Comment by goir** [comment]
#1003 might be related. This issue persists even when removing store_sync or setting to a lower value.

**Comment by Darvame** [comment]
can get something like this only when working with cache before fork (during uwsgi start and etc)

**Comment by Darvame** [comment]
do you have same problem with lazy=1 option?

**Comment by goir** [comment]
It is running with master. Haven't tried lazy, will do this tomorrow.
Still trying to find a way to reliably reproduce this, but no luck so far.

**Comment by Darvame** [comment]
Or maybe this d2f9f7bd3d588172380d40d04e91b756bbddf1f0 from #1169

**Comment by goir** [comment]
I've got a corrupted file from the live servers now. If i start a simple local instance using this cache and trying to write into it I get cache is full. Also all the items have a ttl of 1 hour which is way past now. Those items are not getting cleared. The same on the live server, i've removed it from the load balancer, some caches stay at a fixed number of items. The ttl seems not to work.

I could provide a corrupted cache file and a little test app to help track this down if this helps and someone could debug uwsgi.

**Comment by Darvame** [comment]
sweeper thread does not check items TTL during server start/restart
But lazy mode will solve it (lazy=1)
Also you can run uwsgi.cache_keys(), to read all "alive" data
i/o error also possible

**Comment by goir** [comment]
`lazy` should not be used anymore, i've activated `lazy-apps` now. We'll see how it goes over the day.
But this is expected behaviour when using master = 1 ? This should be documented somewhere since the cache items do not expire and might return outdated results in this situation.

**Comment by Darvame** [comment]
not, lazy-apps, I mean lazy=1 or lazy_expire=1 option inside cache2 like:
_cache2 = name=XXXX,items=30000,keysize=57,blocksize=5,store_delete=1,store=%(cache_dir)/XXXX,ignore_full=1,lazy=1_

**Comment by goir** [comment]
Ah ok, cannot find this option in the docs, thats why i assumed lazy/lazy-apps.
What do they do exactly ?

**Comment by Darvame** [comment]
checking item ttl on access

**Comment by goir** [comment]
Well, this seems to work but if items keys are just valid for a period of time and then are not accessed anymore they will never get deleted and the cache will fill up. It works on a few caches but not all. Using the cache sweeper is perfect. But he needs to check the cache on startup and clean the items whose ttl is expired.
I'll remove the store files for now, this should solve it.

So this is working as intended or is this a bug ? My C++ skills are very rusty but i can check if I can find a solution.

**Comment by Darvame** [comment]
yes, looks like  lazy mode is incomleted, you need to access old item to update ttl, and only then you will able to set new keys.
Lazy mod works fine right now, if you only use same keys all time, my bad
I will make some example patch, to clean up the cache during start up

**Comment by Darvame** [comment]
How _about sweep_on_full=1_ cache2 optio

**Comment by Darvame** [comment]
but, sweep_on_full=N removes whole cache, if can't free least one value (I think it's weird)

**Comment by Darvame** [comment]
this one is better: https://github.com/Darvame/uwsgi/tree/remote-cache-fd-leak
fixed sweep_on_full=N "_surprise_"
and add option **sweep_on_init=1** to sweep cache after init (file load)
so all together **lazy=1,sweep_on_full=1,sweep_on_init=1** must work co

**Comment by unbit** [comment]
@Darvame your solutions are all good, but i think @goir is right,  on startup the sweeper thread should cleanup all the expired items

**Comment by Darvame** [comment]
Sweeper thread must run before main apps start to loading then
okay then, we can just replace my  _uwsgi_cache_sweep_on_init_ with _cache_sweeper_free_items
or leave there both fucns, to let user decide

**Comment by Darvame** [comment]
ok, I updated my fix

**Comment by goir** [comment]
Thanks for working in this! But i think this should be the default, there should not be a special option which is always required to make the cache work correctly :)

**Comment by Darvame** [comment]
Already did it here #1169
